### PR TITLE
first pass on the module

### PR DESCRIPTION
**Pull request by paraynaud**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,137 +1,142 @@
...
 ```julia
-  DenseNetNLPModel = KnetNLPModel(DenseNet; size_minibatch=100, data_train=(xtrn, ytrn), data_test=(xtst, ytst)) # define the KnetNLPModel
+DenseNetNLPModel = KnetNLPModel(DenseNet; size_minibatch=100, data_train=(xtrn, ytrn), data_test=(xtst, ytst))
 ```
 
-
-
 ## Tools associated to a KnetNLPModel
-The dimension of the problem $n$:
+The dimension of the problem `w` ∈ ℝⁿ:
```
Comment (dpo) [inline README.md:102]:
```suggestion
The problem dimension n, where `w` ∈ ℝⁿ:
```

File: README.md [hunk 2]
```diff
@@ -1,137 +1,142 @@
...
 ```julia
-DenseNetNLPModel.meta.nvar
-```
-or in a costly way:
-```julia
-length(vector_params(DenseNetNLPModel))
+n = DenseNetNLPModel.meta.nvar
 ```
 
 ### Get the current variables of the network:
```
Comment (dpo) [inline README.md:107]:
```suggestion
### Get the current network weights:
```

File: README.md [hunk 3]
```diff
@@ -1,137 +1,142 @@
...
 ```julia
 w = vector_params(DenseNetNLPModel)
 ```
 
-### Evaluate the loss function (i.e. the objective function) at the point $w$:
+### Evaluate the loss function (i.e. the objective function) at the point `w`:
```
Comment (dpo) [inline README.md:112]:
```suggestion
### Evaluate the loss function (i.e. the objective function) at `w`:
```

File: README.md [hunk 4]
```diff
@@ -1,137 +1,142 @@
...
-### Evaluate the loss function gradient at the point w (ie the gradient):
+### Evaluate the loss gradient at the point w:
 ```julia
 NLPModels.grad!(DenseNetNLPModel, w, g)
 ```
-The result is stored in `g ::  Vector{T}`(of size `DenseNetNLPModel.meta.nvar`)
+The result is stored in `g ::  Vector{T}`(of size `DenseNetNLPModel.meta.nvar`).
 
-The accuracy of the network can be evaluate with:
+The accuracy of the network can be evaluated with:
```
Comment (dpo) [inline README.md:124]:
```suggestion
### Evaluate the network accuracy

The accuracy of the network can be evaluated with:
```

File: README.md [hunk 5]
```diff
@@ -1,137 +1,142 @@
...
 ```julia
 accuracy(DenseNetNLPModel)
 ```
+`accuracy` use the full training dataset.
+That way, the accuracy will not fluctuate depending on the testing minibatch used.
```
Comment (dpo) [inline README.md:129]:
```suggestion
That way, the accuracy will not fluctuate with the minibatch.
```

File: README.md [hunk 6]
```diff
@@ -1,137 +1,142 @@
...
 
-## Default behaviour
-By default neither the training or the testing minibatch that evaluates the neural network change between evaluations.
-To change the training/testing minibatch use:
-
+## Default behavior
+By default, the training minibatch that evaluates the neural network doesn't change between evaluations.
+To change the training minibatch use:
```
Comment (dpo) [inline README.md:133]:
```suggestion
To change the training minibatch, use:
```

File: README.md [hunk 7]
```diff
@@ -1,137 +1,142 @@
...
-```
-The size of the minibatch will be about the size define previously (may be improved in the future to be dynamic).
-
-This module can be installed with the following command:
-``` julia
-] add https://github.com/paraynaud/KnetNLPModels.jl.git
 ```
+The size of the new minibatch is the size define previously.
 
+The size of the training and testing minibatch can be set to `1/denominator` the size of the dataset with:
```
Comment (dpo) [inline README.md:139]:
```suggestion
The size of the training and test minibatches can be set to `1/p` the size of the dataset with:
```

File: README.md [hunk 8]
```diff
@@ -1,137 +1,142 @@
...
+```julia
+set_size_minibatch!(DenseNetNLPModel, denominator) # denominator::Int > 1
```
Comment (dpo) [inline README.md:141]:
```suggestion
set_size_minibatch!(DenseNetNLPModel, p) # p::Int > 1
```

File: docs/src/tutorial.md [hunk 9]
```diff
@@ -1,118 +1,118 @@
 # KnetNLPModels.jl Tutorial
```
Comment (dpo) [inline docs/src/tutorial.md:1]: Same changes here.

**Review by dpo — CHANGES_REQUESTED** [review]

File: README.md [hunk 10]
```diff
@@ -1,137 +1,142 @@
...
-The length of the vector w must be `DenseNetNLPModel.meta.nvar`
+The length of the vector `w` must be `DenseNetNLPModel.meta.nvar`.
 
-### Evaluate the loss function gradient at the point w (ie the gradient):
+### Evaluate the loss gradient at the point w:
 ```julia
 NLPModels.grad!(DenseNetNLPModel, w, g)
 ```
-The result is stored in `g ::  Vector{T}`(of size `DenseNetNLPModel.meta.nvar`)
+The result is stored in `g ::  Vector{T}`(of size `DenseNetNLPModel.meta.nvar`).
```
[the remark this answers is not in the document]
Comment (paraynaud) [reply README.md:122]: `g` est un `similar(w)`

[the remark this answers is not in the document]
Comment (dpo) [reply README.md:130]: Précise-le.

File: README.md [hunk 11]
```diff
@@ -1,137 +1,151 @@
...
+julia> ] add https://github.com/paraynaud/KnetNLPModels.jl.git
+pkg> test KnetNLPModels
+```
 
-KnetNLPModels is an interface between [Knet.jl](https://github.com/denizyuret/Knet.jl.git)'s classification neural networks and [NLPModels.jl](https://github.com/JuliaSmoothOptimizers/NLPModels.jl.git).
+## Synopsis
+
+A `KnetNLPModel` gives the user access to:
+- the values of the neural network variables/weights `w`;
+- the value of the objective/loss function `L(X, Y; w)` at `w` for a given minibatch `(X,Y)`
```
Comment (dpo) [inline README.md:31]:
```suggestion
- the value of the objective/loss function `L(X, Y; w)` at `w` for a given minibatch `(X,Y)`;
```

File: README.md [hunk 12]
```diff
@@ -1,137 +1,151 @@
...
+- the gradient `∇L(X, Y; w)` of the objective/loss function at `w` for a given mini-batch `(X,Y)`
```
Comment (dpo) [inline README.md:32]:
```suggestion
- the gradient `∇L(X, Y; w)` of the objective/loss function at `w` for a given mini-batch `(X,Y)`.
```

Comment (dpo) [reply README.md:32]: I already suggested the changes above.

File: README.md [hunk 13]
```diff
@@ -1,137 +1,151 @@
...
+- measure the neural network's accuracy at the current `w`.
 
-In addition it provides tools to:
-- Switch the minibatch used to evaluate the neural network
-- Measure the neural network's accuracy at the current point for a given testing mini-batch
+## Example
 
-## Define the layers of interest
-The following code define a dense layer as an evaluable julia structure.
+This step-by-step example suppose prior knowledge [julia](https://julialang.org/) and [Knet.jl](https://github.com/denizyuret/Knet.jl.git).
```
Comment (dpo) [inline README.md:41]:
```suggestion
This step-by-step example assumes prior knowledge of [julia](https://julialang.org/) and [Knet.jl](https://github.com/denizyuret/Knet.jl.git).
```
Also suggested before.

**Review by dpo — CHANGES_REQUESTED** [review]
You skipped a few of my comments. Please also apply them to the tutorial.

**Comment by paraynaud** [comment]
Sorry, I did not see those ones.

**Comment by paraynaud** [comment]
I made the PR #16 from this PR.
It incomporates the last remarks, and the CPU/GPU support that Farhad and I did.

**Comment by dpo** [comment]
> I made the PR https://github.com/paraynaud/KnetNLPModels.jl/pull/16 from this PR.
It incomporates the last remarks,

Please don't do that. Keep PRs orthogonal. We're ready to merge this one.

**Review by dpo — APPROVED** [review]
